README.md
# string-formatting

[![CircleCI](https://circleci.com/gh/suddi/string-formatting.svg?style=svg)](https://circleci.com/gh/suddi/string-formatting)
[![codecov](https://codecov.io/gh/suddi/string-formatting/branch/master/graph/badge.svg)](https://codecov.io/gh/suddi/string-formatting)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/9e6b6ea7bc1a40e9ab11ab621949a83e)](https://www.codacy.com/app/Suddi/string-formatting)
[![npm](https://img.shields.io/npm/v/string-formatting.svg)](https://www.npmjs.com/package/string-formatting)
[![npm](https://img.shields.io/npm/dt/string-formatting.svg)](https://www.npmjs.com/package/string-formatting)
[![David](https://img.shields.io/david/suddi/string-formatting.svg)](https://david-dm.org/suddi/string-formatting)
[![David](https://img.shields.io/david/dev/suddi/string-formatting.svg)](https://david-dm.org/suddi/string-formatting?type=dev)
[![license](https://img.shields.io/github/license/suddi/string-formatting.svg)](https://raw.githubusercontent.com/suddi/string-formatting/master/LICENSE)

[![coverage](https://codecov.io/gh/suddi/string-formatting/branch/master/graphs/commits.svg)](https://codecov.io/gh/suddi/string-formatting)

String formatting module for string beautification, splits string optimally over multiple lines.
Can be used for formatting addresses.

## Installation

````
npm install --save string-formatting
````

## API

````js
const StringFormatter = require('string-formatting');

StringFormatting.apply(<string-to-be-formatted>, <options>);
````

Default configuration, can be overriden with user-defined options:
````js
{
    numLines: 1,                            // number of lines to format to (numLines and lengthOfLine have overlapping use, please see below)
    lengthOfLine: [255],                    // length for each line as a number or length for each specific line in an Array (numLines and lengthOfLine have overlapping use, please see below)
    firstLineRequired: true,                // whether the first line in the array must have a value, else fail
    splitTokenRegex: / /,                   // regex to split the string with
    mergeToken: ' '                         // string to merge the string with in case multiple words join on the same line
};
````

**NOTE:** When both `numLines` defined and `lengthOfLine` is defined as an Array (where each line can have multiple lengths).
`string-formatting` requires that the `numLines` and the length of `lengthOfLine` Array be the same.
This is because in the scenario where `lengthOfLine` is defined per line, `numLines` is an extraneous value and can be omitted.

## Usage

````js
const StringFormatter = require('string-formatting');

const output = StringFormatting.apply('Hello World!', {
    lengthOfLine: [5, 6]                    // the first line is allowed to have a maximum length of 5, the second line, a maximum length of 6
});
console.log(output);
// ['Hello', 'World!']

const output = StringFormatting.apply('Hello World! I am Node.js', {
    numLines: 2,
    lengthOfLine: 12
});
console.log(output);
// ['Hello World!', 'I am Node.js']

const output = StringFormatting.apply('Hello World!', {
    lengthOfLine: [2, 100],
    firstLineRequired: false
});
console.log(output);
// ['', 'Hello World!']

const output = StringFormatting.apply('Hello World!', {
    lengthOfLine: [4, 6],
    splitTokenRegex: /[aeiou]/,
    mergeToken: ';'
});
console.log(output);
// ['H;ll', 'W;rld!']
````

For more workable examples, please see [fixtures](test/fixtures).



CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.0.1] - 2020-04-07
### Added
- Added `.codacy.yml` for Codacy analysis
- Added `.mocharc.yml`

### Changed
- Migrated to using `nyc` from `istanbul`
- Migrated `husky` pre-push hooks
- Migrated to CircleCI v2 pipelines
- Updated `devDependencies` to latest versions
- Updated `package.json` to new Node.js and NPM version compatibility


test/index.js
'use strict';

const fs = require('fs');
const path = require('path');
const expect = require('chai').expect;

const StringFormatter = require('../lib');

function getFixturesPath(filename) {
    const directory = 'fixtures';
    if (filename) {
        return path.join(__dirname, directory, filename);
    }
    return path.join(__dirname, directory);
}

function runTest(inputString, options) {
    try {
        return StringFormatter.apply(inputString, options);
    } catch (error) {
        return error;
    }
}

describe('Integration Tests for string-formatting', function () {
    const filenames = fs.readdirSync(getFixturesPath()).filter(function (filename) {
        return filename.endsWith('.js');
    });

    filenames.map(function (filename, index) {
        it(`CASE ${index + 1}: Testing ${filename}`, function () {
            const T = require(getFixturesPath(filename));
            const input = T.getInput();
            const output = runTest(input.inputString, input.opts);

            if (output instanceof Error) {
                expect(output.message).to.eql(T.getOutput().message);
            } else {
                expect(output).to.deep.eql(T.getOutput());
            }
        });
        return filename;
    });
});


test/utils.js
'use strict';

const expect = require('chai').expect;
const rewire = require('rewire');

const Utils = rewire('../lib/utils');

function getNum(value) {
    return Utils.__get__('getNum')(value);
}

describe('Unit Tests for lib/utils.js', function () {
    context('Testing Utils.getNum', function () {
        it('CASE 1: response for string should be the string length', function () {
            const input = 'Hello World!';
            const expectedResult = input.length;

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 2: response for Object should be the length', function () {
            const input = {answer: 42};
            const expectedResult = input.length;

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 3: response for Array should be the length', function () {
            const input = [4, 5, 6];
            const expectedResult = input.length;

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 4: response for integer should be the value', function () {
            const input = 4;
            const expectedResult = input;

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 5: response for float should be the value', function () {
            const input = 4.2;
            const expectedResult = input;

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 6: response for boolean `true` should be 1', function () {
            const input = true;
            const expectedResult = Number(input);

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 7: response for boolean `false` should be 0', function () {
            const input = false;
            const expectedResult = Number(input);

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 8: response for Function should be 0', function () {
            const input = function () {
                return 42;
            };
            const expectedResult = 0;

            const output = getNum(input);

            expect(output).to.be.eql(expectedResult);
        });
    });

    context('Testing Utils.sumArray', function () {
        it('CASE 1: sumArray works as expected', function () {
            const input = [3, 4, 5];
            const expectedResult = input[0] + input[1] + input[2];

            const output = Utils.sumArray(input);

            expect(output).to.be.eql(expectedResult);
        });
    });
});


test/validator.js
'use strict';

const expect = require('chai').expect;
const rewire = require('rewire');

const Validator = rewire('../lib/validator');

function arrayOf(fieldname, value, func) {
    try {
        return Validator.__get__('arrayOf')(fieldname, value, func);
    } catch (error) {
        return error;
    }
}

function isRequired(fieldname, value) {
    try {
        return Validator.__get__('isRequired')(fieldname, value);
    } catch (error) {
        return error;
    }
}

describe('Unit Tests for lib/validator.js', function () {
    context('Testing Validator.arrayOf', function () {
        it('CASE 1: arrayOf tests for array correctly', function () {
            const fieldname = 'test';
            const value = [3, 4, 5];

            const expectedResult = true;
            const output = arrayOf(fieldname, value);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 2: arrayOf throws error when value is not Array', function () {
            const fieldname = 'test';
            const value = 42;

            const expectedResult = new Error(Validator.getMessage(`${fieldname} must be of Array type`));
            const output = arrayOf(fieldname, value);

            expect(output.message).to.be.eql(expectedResult.message);
        });
    });

    context('Testing Validator.isRequired', function () {
        it('CASE 1: isRequired tests correctly when value exists', function () {
            const fieldname = 'test';
            const value = 42;

            const expectedResult = true;
            const output = isRequired(fieldname, value);

            expect(output).to.be.eql(expectedResult);
        });

        it('CASE 2: isRequired throws error when value does not exist', function () {
            const fieldname = 'test';

            const expectedResult = new Error(Validator.getMessage(`${fieldname} is required`));
            const output = isRequired(fieldname);

            expect(output.message).to.be.eql(expectedResult.message);
        });
    });
});


lib/index.js
'use strict';

const Input = require('./input');
const Options = require('./options');
const Utils = require('./utils');

/**
 * Prepare indexed objects used throughout the formatter
 * @param  {Array} values - values Array to index
 * @param  {number} index - index, so that there is no object manipulation
 * @return {Object} indexed object
 *
 * @example
 * getIndexedObject([3, 4, 5], 2) // => {values: [3, 4, 5], index: 2}
 */
function getIndexedObject(values, index) {
    return {
        values: values,
        index: index
    };
}

/**
 * A test function for the base case, to be used in the recursive calls
 * @param  {Object} words - indexed words object
 * @param  {Object} lengthOfLines - indexed length of lines object
 * @param  {string} currentLine - currently compiled line
 * @param  {Array} formattedArray - current formatted array
 * @return {Array} final result to return from string-formatting
 *
 * @example
 * testBaseCase({values: [2, 3], index: 0}, {values: [2, 3], index: 0}, '', []) // =>
 * testBaseCase({values: [2, 3], index: 2}, {values: [2, 3], index: 0}, '', [1, 2]) // => [1, 2]
 * testBaseCase({values: [2, 3], index: 2}, {values: [2, 3], index: 0}, 'test', [1, 2]) // => [1, 2, 'test']
 * testBaseCase({values: [2, 3], index: 0}, {values: [2, 3], index: 2}, 'test', [1, 2]) // => []
 */
function testBaseCase(words, lengthOfLines, currentLine, formattedArray) {
    if (words.index === words.values.length) {
        return currentLine ? formattedArray.concat(currentLine) : formattedArray;
    } else if (lengthOfLines.index === lengthOfLines.values.length) {
        return [];
    }
}

/**
 * Handle the case when currentLine is defined, called from formatString
 * @param  {Object} words - indexed words object
 * @param  {Object} lengthOfLines - indexed length of lines object
 * @param  {string} mergeToken - the mergeToken to merge strings with
 * @param  {string} currentLine - currently compiled line
 * @param  {Array} formattedArray - current formatted array
 * @return {Array} final result to return from string-formatting
 */
function handleLineDefined(words, lengthOfLines, mergeToken, currentLine, formattedArray) {
    const currentWord = words.values[words.index].trim();
    const currentWordLength = currentWord.length;
    const currentLength = lengthOfLines.values[lengthOfLines.index];

    const computedLength = currentLine.length + mergeToken.length + currentWordLength;

    if (computedLength < currentLength) {
        return formatString(
            getIndexedObject(words.values, words.index + 1),
            getIndexedObject(lengthOfLines.values, lengthOfLines.index),
            mergeToken,
            `${currentLine}${mergeToken}${currentWord}`,
            formattedArray
        );
    } else if (computedLength > currentLength) {
        return formatString(
            getIndexedObject(words.values, words.index),
            getIndexedObject(lengthOfLines.values, lengthOfLines.index + 1),
            mergeToken,
            '',
            formattedArray.concat(currentLine)
        );
    }

    return formatString(
        getIndexedObject(words.values, words.index + 1),
        getIndexedObject(lengthOfLines.values, lengthOfLines.index + 1),
        mergeToken,
        '',
        formattedArray.concat(`${currentLine}${mergeToken}${currentWord}`)
    );
}

/**
 * Handle the case when currentLine is undefined, called from formatString
 * @param  {Object} words - indexed words object
 * @param  {Object} lengthOfLines - indexed length of lines object
 * @param  {string} mergeToken - the mergeToken to merge strings with
 * @param  {string} currentLine - currently compiled line
 * @param  {Array} formattedArray - current formatted array
 * @return {Array} final result to return from string-formatting
 */
function handleLineUndefined(words, lengthOfLines, mergeToken, currentLine, formattedArray) {
    const currentWord = words.values[words.index].trim();
    const currentWordLength = currentWord.length;
    const currentLength = lengthOfLines.values[lengthOfLines.index];

    if (currentWordLength < currentLength) {
        return formatString(
            getIndexedObject(words.values, words.index + 1),
            getIndexedObject(lengthOfLines.values, lengthOfLines.index),
            mergeToken,
            currentWord,
            formattedArray
        );
    } else if (currentWordLength > currentLength) {
        return formatString(
            getIndexedObject(words.values, words.index),
            getIndexedObject(lengthOfLines.values, lengthOfLines.index + 1),
            mergeToken,
            '',
            formattedArray.concat('')
        );
    }

    return formatString(
        getIndexedObject(words.values, words.index + 1),
        getIndexedObject(lengthOfLines.values, lengthOfLines.index + 1),
        mergeToken,
        '',
        formattedArray.concat(currentWord)
    );
}

/**
 * Format the strings into an array
 * @param  {Object} words - indexed words object
 * @param  {Object} lengthOfLines - indexed length of lines object
 * @param  {string} mergeToken - the mergeToken to merge strings with
 * @param  {string} currentLine - currently compiled line
 * @param  {Array} formattedArray - current formatted array
 * @return {Array} final result to return from string-formatting
 */
function formatString(words, lengthOfLines, mergeToken, currentLine, formattedArray) {
    const finalValue = testBaseCase(words, lengthOfLines, currentLine, formattedArray);
    if (finalValue) {
        return finalValue;
    }

    if (currentLine) {
        return handleLineDefined(words, lengthOfLines, mergeToken, currentLine, formattedArray);
    }
    return handleLineUndefined(words, lengthOfLines, mergeToken, currentLine, formattedArray);
}

/**
 * Main function for string-formatting
 * @param  {string} inputString - input string to format
 * @param  {Object} opts - options for string formatting
 * @return {Array} formatted strings in an Array
 *
 * @example
 * apply('Hello World!', {lengthOfLine: [5, 6]}) // => ['Hello', 'World!']
 * apply('Hello World! I am Node.js', {numLines: 2, lengthOfLine: 12}) // => ['Hello World!', 'I am Node.js']
 * apply('Hello World!', {lengthOfLine: [2, 100], firstLineRequired: false}) // => ['', 'Hello World!']
 * apply('Hello World!', {lengthOfLine: [4, 6], splitTokenRegex: /[aeiou]/, mergeToken: ';'}) // => ['H;ll', 'W;rld!']
 */
function apply(inputString, opts) {
    const input = Input.get(inputString);
    const options = Options.get(opts);
    const maximumLength = Utils.sumArray(options.lengthOfLine);

    if (input.length <= options.lengthOfLine[0]) {
        return [input];
    }

    const splitInput = input.split(options.splitTokenRegex);
    if (Utils.sumArray(splitInput) > maximumLength) {
        return [];
    }
    if (options.firstLineRequired &&
        splitInput[0].length > options.lengthOfLine[0]) {
        return [];
    }

    return formatString(
        getIndexedObject(splitInput, 0),
        getIndexedObject(options.lengthOfLine, 0),
        options.mergeToken,
        '',
        []
    );
}

module.exports = {
    apply: apply
};


lib/utils.js
'use strict';

/**
 * Get numeric to compute for different types
 * @param  {*} value - value to compute numeric for
 * @return {number} computed numeric for the value
 *
 * @example
 * getNum('Hello World!') // => 12
 * getNum([3, 4, 5]) // => 3
 * getNum(-1) // => -1
 * getNum(true) // => 1
 * getNum(false) // => 0
 * getNum(function () {}) // => 0
 */
function getNum(value) {
    switch (typeof value) {
        case 'string':
        case 'object':
            return value.length;
        case 'number':
            return value;
        case 'boolean':
            return Number(value);
        default:
            return 0;
    }
}

/**
 * Sum an array of various types
 * @param  {Array} value - array of various types of values
 * @return {number} summed up values from array
 *
 * @example
 * sumArray(['Hello', 'World']) // => 10
 * sumArray([[3, 4, 5], [1, 2]]) // => 5
 * sumArray([3, 1, -1, 4]) // => 7
 * sumArray([true, false, true]) // => 2
 */
function sumArray(value) {
    return value.reduce(function (prev, v) {
        return prev + getNum(v);
    }, 0);
}

module.exports = {
    sumArray: sumArray
};


lib/validator.js
'use strict';

/**
 * Get message for throwing validation errors
 * @param  {*} message - error message
 * @return {string} error message with prefix added
 *
 * @example
 * getMessage('Hello World!') // => '[string-formatting] Hello World!'
 * getMessage(42) // => '[string-formatting] 42'
 */
function getMessage(message) {
    return `[string-formatting] ${message}`;
}

/**
 * Tests whether a value is set or not
 * @param  {*} value - value to validate
 * @return {Boolean} whether validation passes
 *
 * @example
 * isSet() // => false
 * isSet(null) // => false
 * isSet('') // => true
 * isSet('hello') // => true
 * isSet(0) // => true
 * isSet(1) // => true
 */
function isSet(value) {
    return typeof value !== 'undefined' && value !== null;
}

/**
 * Wraps a function
 * @param  {Function} func - function to be wrapped
 * @return {Function} wrapped function
 *
 * @example
 * wrap(function (a, b) { return a + b; }).call({fieldname: 4, value: 5}) // => {fieldname: 4, value: 5}
 */
function wrap(func) {
    return function () {
        const args = Array.prototype.slice.call(arguments);
        func.apply(null, [this.fieldname, this.value].concat(args));
        return this;
    };
}

/**
 * Initializer function for validation
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {boolean} value - value to be validated
 * @return {Object} initialized validator functions
 */
function init(fieldname, value) {
    return {
        fieldname: fieldname,
        value: value,

        // Validation functions
        isBoolean: wrap(isBoolean),
        isInt: wrap(isInt),
        isRegex: wrap(isRegex),
        isStr: wrap(isStr),

        // Array checker
        arrayOf: wrap(arrayOf),

        isRequired: wrap(isRequired)
    };
}

/**
 * Validate whether a value is a Boolean
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {boolean} value - value to validate
 * @return {boolean} whether validation passes
 *
 * @example
 * isBoolean('a', true) // => true
 * isBoolean('a', false) // => true
 * isBoolean() // => true
 */
function isBoolean(fieldname, value) {
    if (isSet(value) && typeof value !== 'boolean') {
        throw new Error(getMessage(`${fieldname} must be of Boolean type`));
    }
    return true;
}

/**
 * Validate whether a value is an Integer
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {boolean} value - value to validate
 * @return {boolean} whether validation passes
 *
 * @example
 * isInt('a', 0) // => true
 * isInt('a', -1) // => true
 * isInt('a', 1) // => true
 * isInt() // => true
 */
function isInt(fieldname, value) {
    if (isSet(value) && !Number.isInteger(value)) {
        throw new Error(getMessage(`${fieldname} must be of Integer type`));
    }
    return true;
}

/**
 * Validate whether a value is a RegExp
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {RegExp} value - value to validate
 * @return {boolean} whether validation passes
 *
 * @example
 * isRegex('a', //) // => true
 * isRegex('a', /, /) // => true
 * isRegex() // => true
 */
function isRegex(fieldname, value) {
    if (isSet(value) && !(value instanceof RegExp)) {
        throw new Error(getMessage(`${fieldname} must be of RegExp type`));
    }
    return true;
}

/**
 * Validate whether a value is a String
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {boolean} value - value to validate
 * @return {boolean} whether validation passes
 *
 * @example
 * isStr('a', 'hello') // => true
 * isStr('a', '') // => true
 * isStr() // => true
 */
function isStr(fieldname, value) {
    if (isSet(value) && typeof value !== 'string') {
        throw new Error(getMessage(`${fieldname} must be of String type`));
    }
    return true;
}

/**
 * Validate whether a value is an Array, as well as its elements
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {boolean} value - value to validate
 * @param  {Function} func - validator function for Array elements
 * @return {boolean} whether validation passes
 *
 * @example
 * arrayOf('a', []) // => true
 * arrayOf('a', [3, 4]) // => true
 * arrayOf('a', [3, 3], function (v) { return v === 3; }) // => true
 * arrayOf() // => true
 */
function arrayOf(fieldname, value, func) {
    if (isSet(value) && !Array.isArray(value)) {
        throw new Error(getMessage(`${fieldname} must be of Array type`));
    } else if (Array.isArray(value) && typeof func === 'function') {
        value.map(func);
    }

    return true;
}

/**
 * Validate whether a value is set
 * @param  {string} fieldname - fieldname to respond in error
 * @param  {boolean} value - value to validate
 * @return {boolean} whether validation passes
 *
 * @example
 * isRequired('a', []) // => true
 * isRequired('a', '') // => true
 * isRequired('a', 'Hello World!') // => true
 * isRequired('a', 0) // => true
 * isRequired('a', 42.1) // => true
 * isRequired('a', true) // => true
 * isRequired('a', false) // => true
 */
function isRequired(fieldname, value) {
    if (!isSet(value)) {
        throw new Error(getMessage(`${fieldname} is required`));
    }
    return true;
}

module.exports = {
    getMessage: getMessage,
    init: init
};


lib/options.js
'use strict';

const Validator = require('./validator');

/**
 * A function to convert to Array if input value is not already an Array
 * @param  {*} value - input value that may or may not be an Array
 * @return {Array} Array output
 *
 * @example
 * convertArray(5) // => [5]
 * convertArray('hello') // => ['hello']
 * convertArray([4, 5, 6]) // => [4, 5, 6]
 * convertArray(['hello']) // ['hello']
 */
function convertArray(value) {
    return Array.isArray(value) ? value : [value];
}

/**
 * Handle numLines and lengthOfLine array
 * @param  {number} numLines - number of lines to create
 * @param  {number} numFromArray - number of lines to create from lengthOfLine
 * @return {number} validated number of lines to create
 *
 * @example
 * getLines(4, 4) // => 4
 * getLines(undefined, 4) // => 4
 * getLines('Hello World!', 4) // => 4
 * getLines(4.2, 4) // => 4
 */
function getLines(numLines, numFromArray) {
    if (Number.isInteger(numLines)) {
        if (numFromArray > 1 && numFromArray !== numLines) {
            throw new Error(Validator.getMessage('numLines and lengthOfLine does not match'));
        } else if (numLines >= 1) {
            return numLines;
        }
    }
    return numFromArray;
}

/**
 * Initialize and validate numLines
 * @param  {number} numLines - number of lines to create
 * @param  {Array|number} lengthOfLine - length of each line
 * @return {number} initialized and validated numLines value
 *
 * @example
 * getNumLines(10) // => 10
 * getNumLines(0) // => 1
 * getNumLines('hello') // => 1
 * getNumLines() // => 1
 * getNumLines(1, [255]) // => 1
 * getNumLines(4, [10, 10, 10, 10]) // => 4
 */
function getNumLines(numLines, lengthOfLine) {
    const value = getLines(
        numLines,
        Array.isArray(lengthOfLine) ? lengthOfLine.length : 1
    );
    return Validator
        .init('numLines', value)
        .isInt()
        .isRequired()
        .value;
}

/**
 * Initialize and validate lengthOfLine
 * @param  {number} numLines - number of lines to create
 * @param  {Array|number} lengthOfLine - length of each line
 * @return {Array} initialized and validated lengthOfLine value
 *
 * @example
 * getLengthOfLine(1, 100) // => [100]
 * getLengthOfLine(4, 100) // => [100, 100, 100, 100]
 * getLengthOfLine(undefined, 0) // => [0]
 * getLengthOfLine(2, [20, 10]) // => [20, 10]
 * getLengthOfLine() // => [255]
 */
function getLengthOfLine(numLines, lengthOfLine) {
    const lines = getLines(
        numLines,
        Array.isArray(lengthOfLine) ? lengthOfLine.length : 1
    );

    const value = lengthOfLine || lengthOfLine === 0 ? lengthOfLine : 255;
    const fillValue = function (v, l) {
        if (v.length === l) {
            return v;
        }
        return Array
            .apply(null, Array(lines))
            .map(function (returnValue) {
                return returnValue;
            }.bind(null, v[0]));
    };

    const filledValue = fillValue(convertArray(value), lines);
    return Validator
        .init('lengthOfLine', filledValue)
        .arrayOf(function (v) {
            return Validator
                .init('Elements of lengthOfLine', v)
                .isInt()
                .isRequired()
                .value;
        })
        .isRequired()
        .value;
}

/**
 * Initialize and validate firstLineRequired
 * @param  {boolean} firstLineRequired - whether first line is required
 * @return {boolean} initialized and validated firstLineRequired value
 *
 * @example
 * getFirstLineRequired(true) // => true
 * getFirstLineRequired(false) // => false
 * getFirstLineRequired() // => true
 */
function getFirstLineRequired(firstLineRequired) {
    const value = firstLineRequired || firstLineRequired === false ? firstLineRequired : true;
    return Validator
        .init('firstLineRequired', value)
        .isBoolean()
        .isRequired()
        .value;
}

/**
 * Initialize and validate splitTokenRegex
 * @param  {string} splitTokenRegex - regex for splitting inputString
 * @return {RegExp} initialized and validated splitTokenRegex
 *
 * @example
 * getSplitTokenRegex(/, /) // => new RegExp(', ')
 * getSplitTokenRegex(//g) // => new RegExp('', 'g')
 * getSplitTokenRegex() // => new RegExp(' ')
 */
function getSplitTokenRegex(splitTokenRegex) {
    return Validator
        .init('splitTokenRegex', splitTokenRegex ? splitTokenRegex : / /)
        .isRegex()
        .isRequired()
        .value;
}

/**
 * Initialize and validate mergeToken
 * @param  {string} mergeToken - string for merging inputString
 * @return {string} initialized and validated mergeToken
 *
 * @example
 * getMergeToken(' ') // => ' '
 * getMergeToken() // => ' '
 */
function getMergeToken(mergeToken) {
    return Validator
        .init('mergeToken', mergeToken || mergeToken === '' ? mergeToken : ' ')
        .isStr()
        .isRequired()
        .value;
}

/**
 * Retrieve all filled options set with default values as needed
 * @param  {Object} options - options for string-formatting lib
 * @return {Object} filled options
 *
 * @example
 * get({numLines: 0, lengthOfLine: 0, firstLineRequired: false, splitTokenRegex: /, /, mergeToken: 'x'})
 * // => {numLines: 1, lengthOfLine: [0], firstLineRequired: false, splitTokenRegex: new RegExp(', '), mergeToken: 'x'}
 * get({}) // => {numLines: 1, lengthOfLine: [255], firstLineRequired: true, splitTokenRegex: new RegExp(' '), mergeToken: ' '}
 */
function get(options) {
    return {
        numLines: getNumLines(options.numLines, options.lengthOfLine),
        lengthOfLine: getLengthOfLine(options.numLines, options.lengthOfLine),
        firstLineRequired: getFirstLineRequired(options.firstLineRequired),
        splitTokenRegex: getSplitTokenRegex(options.splitTokenRegex),
        mergeToken: getMergeToken(options.mergeToken)
    };
}

module.exports = {
    get: get
};
